### ISSUE TITLE
Filtering entities using labels

### ISSUE THREAD

**Issue by kschiffer** [TheThingsNetwork/lorawan-stack#5710]
### Summary

As a means of logically grouping entities and assigning different treatments or performing batch actions based on such groups, we should introduce entity tags.

### Current Situation

There is no way of grouping entities other than maybe assigning IDs with a prefix, which is only an implicit way of grouping.

### Why do we need this? Who uses it, and when?

As stated above, tags are a way of grouping entities. Such grouping has several benefits:

- Enabling explicit visual grouping as a means of organizing an entity fleet logically and facilitate device fleet management.
- Improving entity search.
- Assigning different settings to groups of end devices based on tags. For e.g. we could make integrations assignable by tags so that multiple integrations can be set up for different slices of an application's end devices.
- Selecting devices for batch operations, such as moving devices between applications or deleting end devices in batch.

### Proposed Implementation

This issue should be scoped to only the tags themselves and their assignment to entities.

We would need a new tag model which contains information about the tag such as:
- ID
- Name
- Description
- Color

Then we need to add an array field to the entities `tags` which contains the list of assigned tag IDs.

To assign the tags, we would require an API that allows adding/removing a list of tags to/from a list of entities.

Then eventually, we need to update our list/search APIs to include filtering by tag IDs.

@adriansmares @johanstokking assigning you initially to provide some input and to reassign as appropriate.

### Contributing

- [X] I can help by doing more research.
- [X] I can help by implementing the feature after the proposal above is approved.
- [X] I can help by testing the feature before it's released.

### Code of Conduct

- [X] I agree to follow TTN's [Community Code of Conduct](https://github.com/TheThingsNetwork/Manifesto/blob/master/CodeOfConduct.md).

**Comment by johanstokking** [comment]
As discussed offline, this would work very much like GitHub labels. This would indeed not just be for end devices, but for other entities as well, most notably gateways and applications.

This is different from attributes in the sense that this is a list of independent values, and we should efficiently filter on the presence of a value. Since TTS would have first class support for this, it wouldn't be an attribute anyway, as those are custom user defined things.

**Comment by KrishnaIyer** [comment]
@pierrephz for the UX.

**Comment by pierrephz** [comment]
Here is how the label could work on the console:

# Label page
- List of labels with `Name`, `Description`, `Related entities` (how many applications, gateways and end devices have this label), and `Actions` (Edit and Delete)

<img width="1458" alt="Image" src="https://github.com/user-attachments/assets/6cbd5cdb-df5e-44e0-8c5a-7274bbbb27cf" />

<img width="1466" alt="Image" src="https://github.com/user-attachments/assets/0088cdef-36f6-4fd7-8fd0-f0add748c730" />

## How to create a label?
- From the label page, same interaction like on Github; Click on `Create label`, it opens a label creation panel.
- From the entity creation page, add a `End device propriety` section to allow the user to assign or create a label from there.

<img width="1606" alt="Image" src="https://github.com/user-attachments/assets/04687684-4e8a-454c-8c71-78879b4e59af" />

<img width="1604" alt="Image" src="https://github.com/user-attachments/assets/78c5c174-21de-49e2-ab52-88b0d4d5fe70" />

## How to access and use a label?

### From the Home page
- Add a label interaction in the Header.
- Like with the Notification interaction; display the most recent labels with the number of related entities.
- Add a link at the bottom of the dropdown to the label page to see all of them.

<img width="1526" alt="Image" src="https://github.com/user-attachments/assets/02c10706-e0c8-43e3-9a16-2231abed2ee9" />

### With the Search interaction
- Add a nested input below the main search input to add an extra filter with label on the search. Having a nested input below the main one allows us to keep a hierarchy and keep the main search input in the center of the attention.
- When opened, the nested label input also have a dropdown to allow the user to select a label from a list, without having to type it. It is an important feature since the user might not know every label by heart.

<img width="1505" alt="Image" src="https://github.com/user-attachments/assets/11cff95c-2932-495a-b3db-7df28c4d431d" />

<img width="1469" alt="Image" src="https://github.com/user-attachments/assets/7947711f-8b9f-425e-b45e-34887749370e" />

<img width="1481" alt="Image" src="https://github.com/user-attachments/assets/aed8637c-0d74-4e04-b8c6-71617ae405da" />

### In a entity page
- The label is visible next to the name of the entity at the top of the page.

<img width="1519" alt="Image" src="https://github.com/user-attachments/assets/9c16b9d9-7ee3-49ec-8844-ed948f86e5f3" />

@johanstokking @KrishnaIyer this is a proposal on how to use the labels in the console, let me know if you have feedback or if you think something is missing, then we can work with @ryaplots on the implementation

**Comment by KrishnaIyer** [comment]
Apart from spelling errors, this generally looks really nice.

I have a couple of questions:
1. This only shows adding labels to new entities. How can users assign them to exisiting ones?
2. For my untrained UX eyes, I'm not so sure of the extra roundedness (that's not a real word). We don't have any other element with this much rounded corners and to my eyes it looks odd.

**Comment by ryaplots** [comment]
>     1. This only shows adding labels to new entities. How can users assign them to exisiting ones?
True! We forgot to discuss how to assign labels to existing entities or how to edit labels. It should be possible to open the overview of any entity and add/edit the labels

**Comment by pierrephz** [comment]
> 2. For my untrained UX eyes, I'm not so sure of the extra roundedness (that's not a real word). We don't have any other element with this much rounded corners and to my eyes it looks odd.

It make sense that we don't have any other element with this roundness, since we are introducing a new feature (labels).
Having them "square" like our button would make them too similar, but they serve distinct purposes (categorisation for the labels / actions for the buttons) so they have to be different.

I was using the same UI as Github there, but if you think they are too catchy with the background colors, I can look into a more subtle UI, like color dot + text and color borders

**Comment by ryaplots** [comment]
We also need, in each entity list, a way to filter by label. @pierrephz

**Comment by pierrephz** [comment]
> 1. This only shows adding labels to new entities. How can users assign them to exisiting ones?

Here is a proposal to add assign a label to en existing entity (from the entity page):
- Add a clickable label placeholder (+ Add label) in the labels' spot next to the entity name. When they are already some labels, add this placeholder at the end of the labels' raw.
- Clicking on it will open a modal where you can add an existing label, or create a new one.

<img width="1452" alt="Image" src="https://github.com/user-attachments/assets/c5062505-fa1a-4842-96f9-9e32337d801b" />

<img width="1470" alt="Image" src="https://github.com/user-attachments/assets/39be9662-e7d0-4def-9b81-d38ee1359b94" />

<img width="1469" alt="Image" src="https://github.com/user-attachments/assets/83ae786b-f22f-4737-88f4-15e3f11ef9c7" />

**Comment by KrishnaIyer** [comment]
> Having them "square" like our button would make them too similar, but they serve distinct purposes (categorisation for the labels / actions for the buttons) so they have to be different.

Yeah, you're probably right. Let's stick to this for now and see how it works out. Thanks!

**Comment by pierrephz** [comment]
> We also need, in each entity list, a way to filter by label. [@pierrephz](https://github.com/pierrephz)

Here is a way to filter entities by labels:
- Add a `Label` filter next to the search bar
- On click, open a dropdown list where you can search, select labels, and clear the current selection
- If some labels are selected, the filter gets "active" (black) to show user some labels are being applied

<img width="1424" alt="Image" src="https://github.com/user-attachments/assets/776568d3-02ca-4447-a619-e89cab05f125" />

<img width="1536" alt="Image" src="https://github.com/user-attachments/assets/13210b7a-1a43-4c5b-9d38-069106dde0d9" />

<img width="1531" alt="Image" src="https://github.com/user-attachments/assets/eca31d60-7e41-46f9-933e-cb138f539944" />

<img width="1543" alt="Image" src="https://github.com/user-attachments/assets/646853cd-30ce-4206-9c73-37912a73140b" />

Let me know if you have any thoughts

**Comment by johanstokking** [comment]
Just chiming in here that this is gonna be really great.

**Comment by KrishnaIyer** [comment]
Tested in https://github.com/TheThingsIndustries/lorawan-stack-support/issues/1267 and can be closed.
